CultureAnalyzer/middleware.py

```diff
@@ -0,0 +1,74 @@
+from django.contrib.sessions.models import Session
+from django.db import transaction
+
+from CultureAnalyzer.util import login_redirect
+
+__all__ = ['AuthRequiredMiddleware', 'SwitchSessionDataMiddleware']
+
+
+class AuthRequiredMiddleware:
+    def __init__(self, response):
+        self.get_response = response
+
+    @login_redirect
+    def __call__(self, request):
+        return self.get_response(request)
+
+
+class SwitchSessionDataMiddleware:
+    def __init__(self, get_response):
+        self.get_response = get_response
+
+    def __call__(self, request):
+        self._check_session(request)
+
+        response = self.get_response(request)
+
+        return response
+
+    @transaction.atomic()
+    def _check_session(self, request):
+        """
+        Method, which checks if there previous session, that`s mean has changed
+         the client from which user authenticated.
+
+        :var current_session_key(str): hashed key, which identifies current
+         session;
+        :var stored_session_key(str): hashed key, which identifies previous
+         session;
+        """
+        if request.user.is_authenticated:
+            current_session_key = request.session.session_key
+            stored_session_key = request.user.logged_in_user.session_key
+
+            if stored_session_key and stored_session_key != current_session_key:
+                self._switch_session_data(request, stored_session_key)
+            request.user.logged_in_user.session_key = current_session_key
+            request.user.logged_in_user.save()
+
```

Comment:
I'm not sure, but maybe the next lines should be also inside `transaction`
https://github.com/IF-093-Python/CultureAnalyzer/blob/df68f76ca428c68e7e6e28f0648aaf939786920e/CultureAnalyzer/middleware.py#L46-L47

You can to append this part of code to `_switch_session_data` method